eradiate/radprops/particles.py

```diff
@@ -0,0 +1,386 @@
+"""
+Particle layers.
+"""
+from __future__ import annotations
+
+from typing import Union
+
+import attr
+import numpy as np
+import pint
+import pinttr
+import xarray as xr
+
+from .particle_dist import ParticleDistribution, ParticleDistributionFactory
+from .. import path_resolver
+from .._attrs import documented, parse_docs
+from ..contexts import SpectralContext
+from ..units import to_quantity
+from ..units import unit_context_config as ucc
+from ..units import unit_registry as ureg
+from ..validators import is_positive
+
+
+@parse_docs
+@attr.s
+class ParticleLayer:
+    """
+    Particle layer.
+
+    The particle layer has a vertical extension specified by a bottom altitude
+    (set by ``bottom``) and a top altitude (set by ``top``).
+    Inside the layer, the particles number is distributed according to a
+    distribution (set by ``distribution``).
+    See :class:`.ParticleDistributionFactory` for the available distribution
+    types and corresponding parameters.
+    The particle layer is itself divided into a number of (sub-)layers
+    (``n_layers``) to allow to describe the variations of the particles number
+    with altitude.
+    The total number of particles in the layer is adjusted so that the
+    particle layer's optical thickness at 550 nm meet a specified value
+    (``tau_550``).
+    The particles radiative properties are specified by a data set
+    (``dataset``).
+    """
+
+    bottom = documented(
+        pinttr.ib(
+            default=ureg.Quantity(0.0, ureg.km),
+            converter=pinttr.converters.to_units(ucc.deferred("length")),
+            validator=[
+                is_positive,
+                pinttr.validators.has_compatible_units,
+            ],
+            units=ucc.deferred("length"),
+        ),
+        doc="Bottom altitude of the particle layer.\n"
+        "\n"
+        "Unit-enabled field (default: ucc[length])",
+        type="float",
+        default="0 km",
+    )
+
+    top = documented(
+        pinttr.ib(
+            units=ucc.deferred("length"),
+            default=ureg.Quantity(1.0, ureg.km),
+            converter=pinttr.converters.to_units(ucc.deferred("length")),
+            validator=[
+                is_positive,
+                pinttr.validators.has_compatible_units,
+            ],
+        ),
+        doc="Top altitude of the particle layer.\n"
+        "\n"
+        "Unit-enabled field (default: ucc[length]).",
+        type="float",
+        default="1 km.",
+    )
+
+    @bottom.validator
+    @top.validator
+    def _bottom_and_top_validator(instance, attribute, value):
+        if instance.bottom >= instance.top:
+            raise ValueError("bottom altitude must be lower than top altitude")
+
+    distribution = documented(
+        attr.ib(
+            default={"type": "uniform"},
```

Comment:
Style: I'd prefer if you'd use the constructor itself here:
```suggestion
            factory=UniformParticleDistribution,
```